<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Data Preprocessing for Bioinformatics | Introduction to Deep Learning for Bioinformatics</title>
  <meta name="description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="generator" content="bookdown 0.45 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Data Preprocessing for Bioinformatics | Introduction to Deep Learning for Bioinformatics" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://danyukeshagithub.io/DeepLearningFundamentals/assets/images/cover.png" />
  <meta property="og:description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="github-repo" content="danymukesha/DeepLearningFundamentals" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Data Preprocessing for Bioinformatics | Introduction to Deep Learning for Bioinformatics" />
  
  <meta name="twitter:description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="twitter:image" content="https://danyukeshagithub.io/DeepLearningFundamentals/assets/images/cover.png" />

<meta name="author" content="Dany Mukesha" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="implementing-neural-networks-with-keras.html"/>
<link rel="next" href="convolutional-neural-networks-for-genomics.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Intro to DL</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>0.1</b> Course Overview</a>
<ul>
<li class="chapter" data-level="0.1.1" data-path="index.html"><a href="index.html#learning-goals"><i class="fa fa-check"></i><b>0.1.1</b> Learning Goals</a></li>
<li class="chapter" data-level="0.1.2" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>0.1.2</b> Prerequisites</a></li>
<li class="chapter" data-level="0.1.3" data-path="index.html"><a href="index.html#setup-instructions"><i class="fa fa-check"></i><b>0.1.3</b> Setup Instructions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html"><i class="fa fa-check"></i><b>1</b> Introduction to Deep Learning</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#what-is-deep-learning"><i class="fa fa-check"></i><b>1.1</b> What is Deep Learning?</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#key-concepts"><i class="fa fa-check"></i><b>1.1.1</b> Key Concepts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#why-deep-learning-in-bioinformatics"><i class="fa fa-check"></i><b>1.2</b> Why Deep Learning in Bioinformatics?</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#a-simple-biological-example-transcription-factor-binding-prediction"><i class="fa fa-check"></i><b>1.3</b> A Simple Biological Example: Transcription Factor Binding Prediction</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html"><i class="fa fa-check"></i><b>2</b> Implementing Neural Networks with Keras</a>
<ul>
<li class="chapter" data-level="2.1" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#what-is-keras"><i class="fa fa-check"></i><b>2.1</b> What is Keras?</a></li>
<li class="chapter" data-level="2.2" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#your-first-neural-network"><i class="fa fa-check"></i><b>2.2</b> Your First Neural Network</a></li>
<li class="chapter" data-level="2.3" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#model-anatomy"><i class="fa fa-check"></i><b>2.3</b> Model Anatomy</a></li>
<li class="chapter" data-level="2.4" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#training-the-model"><i class="fa fa-check"></i><b>2.4</b> Training the Model</a></li>
<li class="chapter" data-level="2.5" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#evaluating-the-model"><i class="fa fa-check"></i><b>2.5</b> Evaluating the Model</a></li>
<li class="chapter" data-level="2.6" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html"><i class="fa fa-check"></i><b>3</b> Data Preprocessing for Bioinformatics</a>
<ul>
<li class="chapter" data-level="3.1" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#encoding-categorical-data"><i class="fa fa-check"></i><b>3.1</b> Encoding Categorical Data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#one-hot-encoding-for-dna-sequences"><i class="fa fa-check"></i><b>3.1.1</b> One-Hot Encoding for DNA Sequences</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#handling-variable-length-sequences"><i class="fa fa-check"></i><b>3.2</b> Handling Variable-Length Sequences</a></li>
<li class="chapter" data-level="3.3" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#data-generators"><i class="fa fa-check"></i><b>3.3</b> Data Generators</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html"><i class="fa fa-check"></i><b>4</b> Convolutional Neural Networks for Genomics</a>
<ul>
<li class="chapter" data-level="4.1" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#the-convolutional-layer"><i class="fa fa-check"></i><b>4.1</b> The Convolutional Layer</a></li>
<li class="chapter" data-level="4.2" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#case-study-basset"><i class="fa fa-check"></i><b>4.2</b> Case Study: Basset</a></li>
<li class="chapter" data-level="4.3" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#exercise-build-a-tf-binding-predictor"><i class="fa fa-check"></i><b>4.3</b> Exercise: Build a TF Binding Predictor</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html"><i class="fa fa-check"></i><b>5</b> Large Language Models in Bioinformatics</a>
<ul>
<li class="chapter" data-level="5.1" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#from-word-embeddings-to-biological-embeddings"><i class="fa fa-check"></i><b>5.1</b> From Word Embeddings to Biological Embeddings</a></li>
<li class="chapter" data-level="5.2" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#pre-trained-models-protbert-and-dnabert"><i class="fa fa-check"></i><b>5.2</b> Pre-trained Models: ProtBERT and DNABERT</a></li>
<li class="chapter" data-level="5.3" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#how-to-use-a-pre-trained-model"><i class="fa fa-check"></i><b>5.3</b> How to Use a Pre-trained Model</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#example-using-dnabert-for-promoter-prediction"><i class="fa fa-check"></i><b>5.3.1</b> Example: Using DNABERT for Promoter Prediction</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#potential-applications"><i class="fa fa-check"></i><b>5.4</b> Potential Applications</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html"><i class="fa fa-check"></i><b>6</b> Advanced Topics &amp; Best Practices</a>
<ul>
<li class="chapter" data-level="6.1" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#regularization-fighting-overfitting"><i class="fa fa-check"></i><b>6.1</b> Regularization: Fighting Overfitting</a></li>
<li class="chapter" data-level="6.2" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#hyperparameter-tuning"><i class="fa fa-check"></i><b>6.2</b> Hyperparameter Tuning</a></li>
<li class="chapter" data-level="6.3" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#interpretability-what-did-my-model-learn"><i class="fa fa-check"></i><b>6.3</b> Interpretability: What Did My Model Learn?</a></li>
<li class="chapter" data-level="6.4" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#resources-for-the-future"><i class="fa fa-check"></i><b>6.4</b> Resources for the Future</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html"><i class="fa fa-check"></i><b>7</b> References and Further Reading</a>
<ul>
<li class="chapter" data-level="7.1" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#books-and-papers"><i class="fa fa-check"></i><b>7.1</b> Books and Papers</a></li>
<li class="chapter" data-level="7.2" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#online-courses"><i class="fa fa-check"></i><b>7.2</b> Online Courses</a></li>
<li class="chapter" data-level="7.3" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#useful-links"><i class="fa fa-check"></i><b>7.3</b> Useful Links</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Deep Learning for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-preprocessing-for-bioinformatics" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Data Preprocessing for Bioinformatics<a href="data-preprocessing-for-bioinformatics.html#data-preprocessing-for-bioinformatics" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Neural networks require data to be in a numerical, normalized format. Biological data often requires special encoding.</p>
<div id="encoding-categorical-data" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Encoding Categorical Data<a href="data-preprocessing-for-bioinformatics.html#encoding-categorical-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="one-hot-encoding-for-dna-sequences" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> One-Hot Encoding for DNA Sequences<a href="data-preprocessing-for-bioinformatics.html#one-hot-encoding-for-dna-sequences" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This is the most common method for representing sequences. Each nucleotide is converted into a vector.</p>
<ul>
<li>A -&gt; <code>[1, 0, 0, 0]</code></li>
<li>T -&gt; <code>[0, 1, 0, 0]</code></li>
<li>C -&gt; <code>[0, 0, 1, 0]</code></li>
<li>G -&gt; <code>[0, 0, 0, 1]</code></li>
<li>N -&gt; <code>[0, 0, 0, 0]</code></li>
</ul>
<p>A sequence <code>"ATCG"</code> becomes a 2D array of shape <code>(4, 4)</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="data-preprocessing-for-bioinformatics.html#cb4-1" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb4-2"><a href="data-preprocessing-for-bioinformatics.html#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="data-preprocessing-for-bioinformatics.html#cb4-3" tabindex="-1"></a><span class="kw">def</span> one_hot_encode_dna(sequence):</span>
<span id="cb4-4"><a href="data-preprocessing-for-bioinformatics.html#cb4-4" tabindex="-1"></a>    <span class="co">&quot;&quot;&quot;One-hot encode a DNA sequence.&quot;&quot;&quot;</span></span>
<span id="cb4-5"><a href="data-preprocessing-for-bioinformatics.html#cb4-5" tabindex="-1"></a>    mapping <span class="op">=</span> {<span class="st">&#39;A&#39;</span>: [<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>], <span class="st">&#39;T&#39;</span>: [<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>], <span class="st">&#39;C&#39;</span>: [<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">0</span>], <span class="st">&#39;G&#39;</span>: [<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>]}</span>
<span id="cb4-6"><a href="data-preprocessing-for-bioinformatics.html#cb4-6" tabindex="-1"></a>    one_hot <span class="op">=</span> []</span>
<span id="cb4-7"><a href="data-preprocessing-for-bioinformatics.html#cb4-7" tabindex="-1"></a>    <span class="cf">for</span> base <span class="kw">in</span> sequence:</span>
<span id="cb4-8"><a href="data-preprocessing-for-bioinformatics.html#cb4-8" tabindex="-1"></a>        one_hot.append(mapping.get(base, [<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>])) <span class="co"># Handle &#39;N&#39; or other chars</span></span>
<span id="cb4-9"><a href="data-preprocessing-for-bioinformatics.html#cb4-9" tabindex="-1"></a>    <span class="cf">return</span> np.array(one_hot)</span>
<span id="cb4-10"><a href="data-preprocessing-for-bioinformatics.html#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="data-preprocessing-for-bioinformatics.html#cb4-11" tabindex="-1"></a><span class="co"># Example</span></span>
<span id="cb4-12"><a href="data-preprocessing-for-bioinformatics.html#cb4-12" tabindex="-1"></a>seq <span class="op">=</span> <span class="st">&quot;ATCGN&quot;</span></span>
<span id="cb4-13"><a href="data-preprocessing-for-bioinformatics.html#cb4-13" tabindex="-1"></a>encoded_seq <span class="op">=</span> one_hot_encode_dna(seq)</span>
<span id="cb4-14"><a href="data-preprocessing-for-bioinformatics.html#cb4-14" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Sequence: </span><span class="sc">{</span>seq<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb4-15"><a href="data-preprocessing-for-bioinformatics.html#cb4-15" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Encoded shape: </span><span class="sc">{</span>encoded_seq<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb4-16"><a href="data-preprocessing-for-bioinformatics.html#cb4-16" tabindex="-1"></a><span class="bu">print</span>(encoded_seq)</span></code></pre></div>
</div>
</div>
<div id="handling-variable-length-sequences" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Handling Variable-Length Sequences<a href="data-preprocessing-for-bioinformatics.html#handling-variable-length-sequences" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For sequences of different lengths, we use <strong>padding</strong>. We choose a maximum length and pad shorter sequences with zeros.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="data-preprocessing-for-bioinformatics.html#cb5-1" tabindex="-1"></a><span class="im">from</span> tensorflow.keras.preprocessing.sequence <span class="im">import</span> pad_sequences</span>
<span id="cb5-2"><a href="data-preprocessing-for-bioinformatics.html#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="data-preprocessing-for-bioinformatics.html#cb5-3" tabindex="-1"></a><span class="co"># Assume `list_of_encoded_sequences` is a list of one-hot encoded sequences</span></span>
<span id="cb5-4"><a href="data-preprocessing-for-bioinformatics.html#cb5-4" tabindex="-1"></a>padded_sequences <span class="op">=</span> pad_sequences(list_of_encoded_sequences, maxlen<span class="op">=</span><span class="dv">200</span>, padding<span class="op">=</span><span class="st">&#39;post&#39;</span>, dtype<span class="op">=</span><span class="st">&#39;float32&#39;</span>)</span>
<span id="cb5-5"><a href="data-preprocessing-for-bioinformatics.html#cb5-5" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Padded sequences shape: </span><span class="sc">{</span>padded_sequences<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">&quot;</span>) <span class="co"># (num_samples, 200, 4)</span></span></code></pre></div>
</div>
<div id="data-generators" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Data Generators<a href="data-preprocessing-for-bioinformatics.html#data-generators" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For large datasets that don’t fit in memory, we use <code>tf.data.Dataset</code> or Keras’ <code>ImageDataGenerator</code> to stream data from disk in batches.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="data-preprocessing-for-bioinformatics.html#cb6-1" tabindex="-1"></a><span class="co"># Example for reading from text files</span></span>
<span id="cb6-2"><a href="data-preprocessing-for-bioinformatics.html#cb6-2" tabindex="-1"></a><span class="kw">def</span> file_data_generator(file_paths, labels, batch_size<span class="op">=</span><span class="dv">32</span>):</span>
<span id="cb6-3"><a href="data-preprocessing-for-bioinformatics.html#cb6-3" tabindex="-1"></a>    num_samples <span class="op">=</span> <span class="bu">len</span>(file_paths)</span>
<span id="cb6-4"><a href="data-preprocessing-for-bioinformatics.html#cb6-4" tabindex="-1"></a>    <span class="cf">while</span> <span class="va">True</span>: <span class="co"># Loop forever so the generator never terminates</span></span>
<span id="cb6-5"><a href="data-preprocessing-for-bioinformatics.html#cb6-5" tabindex="-1"></a>        <span class="cf">for</span> offset <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, num_samples, batch_size):</span>
<span id="cb6-6"><a href="data-preprocessing-for-bioinformatics.html#cb6-6" tabindex="-1"></a>            batch_paths <span class="op">=</span> file_paths[offset:offset<span class="op">+</span>batch_size]</span>
<span id="cb6-7"><a href="data-preprocessing-for-bioinformatics.html#cb6-7" tabindex="-1"></a>            batch_labels <span class="op">=</span> labels[offset:offset<span class="op">+</span>batch_size]</span>
<span id="cb6-8"><a href="data-preprocessing-for-bioinformatics.html#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="data-preprocessing-for-bioinformatics.html#cb6-9" tabindex="-1"></a>            batch_sequences <span class="op">=</span> []</span>
<span id="cb6-10"><a href="data-preprocessing-for-bioinformatics.html#cb6-10" tabindex="-1"></a>            <span class="cf">for</span> path <span class="kw">in</span> batch_paths:</span>
<span id="cb6-11"><a href="data-preprocessing-for-bioinformatics.html#cb6-11" tabindex="-1"></a>                <span class="cf">with</span> <span class="bu">open</span>(path, <span class="st">&#39;r&#39;</span>) <span class="im">as</span> f:</span>
<span id="cb6-12"><a href="data-preprocessing-for-bioinformatics.html#cb6-12" tabindex="-1"></a>                    seq <span class="op">=</span> f.read().strip()</span>
<span id="cb6-13"><a href="data-preprocessing-for-bioinformatics.html#cb6-13" tabindex="-1"></a>                encoded_seq <span class="op">=</span> one_hot_encode_dna(seq)</span>
<span id="cb6-14"><a href="data-preprocessing-for-bioinformatics.html#cb6-14" tabindex="-1"></a>                batch_sequences.append(encoded_seq)</span>
<span id="cb6-15"><a href="data-preprocessing-for-bioinformatics.html#cb6-15" tabindex="-1"></a></span>
<span id="cb6-16"><a href="data-preprocessing-for-bioinformatics.html#cb6-16" tabindex="-1"></a>            <span class="co"># Pad the sequences in the batch</span></span>
<span id="cb6-17"><a href="data-preprocessing-for-bioinformatics.html#cb6-17" tabindex="-1"></a>            batch_x <span class="op">=</span> pad_sequences(batch_sequences, maxlen<span class="op">=</span><span class="dv">200</span>, padding<span class="op">=</span><span class="st">&#39;post&#39;</span>)</span>
<span id="cb6-18"><a href="data-preprocessing-for-bioinformatics.html#cb6-18" tabindex="-1"></a>            <span class="cf">yield</span> batch_x, batch_labels</span>
<span id="cb6-19"><a href="data-preprocessing-for-bioinformatics.html#cb6-19" tabindex="-1"></a></span>
<span id="cb6-20"><a href="data-preprocessing-for-bioinformatics.html#cb6-20" tabindex="-1"></a><span class="co"># Usage</span></span>
<span id="cb6-21"><a href="data-preprocessing-for-bioinformatics.html#cb6-21" tabindex="-1"></a><span class="co"># train_generator = file_data_generator(train_files, train_labels)</span></span>
<span id="cb6-22"><a href="data-preprocessing-for-bioinformatics.html#cb6-22" tabindex="-1"></a><span class="co"># model.fit(train_generator, epochs=10, steps_per_epoch=len(train_files)//batch_size)</span></span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="implementing-neural-networks-with-keras.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="convolutional-neural-networks-for-genomics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/danymukesha/DeepLearningFundamentals/edit/BRANCH/03-Data-Preprocessing.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["Intro_Deep_Learning_Bioinformatics.pdf", "Intro_Deep_Learning_Bioinformatics.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
