<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Large Language Models in Bioinformatics | Introduction to Deep Learning for Bioinformatics</title>
  <meta name="description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="generator" content="bookdown 0.45 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Large Language Models in Bioinformatics | Introduction to Deep Learning for Bioinformatics" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://danyukeshagithub.io/DeepLearningFundamentals/assets/images/cover.png" />
  <meta property="og:description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="github-repo" content="danymukesha/DeepLearningFundamentals" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Large Language Models in Bioinformatics | Introduction to Deep Learning for Bioinformatics" />
  
  <meta name="twitter:description" content="“Deep Learning fundamentals and their application in bioinformatics, using Keras.”" />
  <meta name="twitter:image" content="https://danyukeshagithub.io/DeepLearningFundamentals/assets/images/cover.png" />

<meta name="author" content="Dany Mukesha" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="convolutional-neural-networks-for-genomics.html"/>
<link rel="next" href="advanced-topics-best-practices.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Intro to DL</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>0.1</b> Course Overview</a>
<ul>
<li class="chapter" data-level="0.1.1" data-path="index.html"><a href="index.html#learning-goals"><i class="fa fa-check"></i><b>0.1.1</b> Learning Goals</a></li>
<li class="chapter" data-level="0.1.2" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>0.1.2</b> Prerequisites</a></li>
<li class="chapter" data-level="0.1.3" data-path="index.html"><a href="index.html#setup-instructions"><i class="fa fa-check"></i><b>0.1.3</b> Setup Instructions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html"><i class="fa fa-check"></i><b>1</b> Introduction to Deep Learning</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#what-is-deep-learning"><i class="fa fa-check"></i><b>1.1</b> What is Deep Learning?</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#key-concepts"><i class="fa fa-check"></i><b>1.1.1</b> Key Concepts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#why-deep-learning-in-bioinformatics"><i class="fa fa-check"></i><b>1.2</b> Why Deep Learning in Bioinformatics?</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-deep-learning.html"><a href="introduction-to-deep-learning.html#a-simple-biological-example-transcription-factor-binding-prediction"><i class="fa fa-check"></i><b>1.3</b> A Simple Biological Example: Transcription Factor Binding Prediction</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html"><i class="fa fa-check"></i><b>2</b> Implementing Neural Networks with Keras</a>
<ul>
<li class="chapter" data-level="2.1" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#what-is-keras"><i class="fa fa-check"></i><b>2.1</b> What is Keras?</a></li>
<li class="chapter" data-level="2.2" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#your-first-neural-network"><i class="fa fa-check"></i><b>2.2</b> Your First Neural Network</a></li>
<li class="chapter" data-level="2.3" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#model-anatomy"><i class="fa fa-check"></i><b>2.3</b> Model Anatomy</a></li>
<li class="chapter" data-level="2.4" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#training-the-model"><i class="fa fa-check"></i><b>2.4</b> Training the Model</a></li>
<li class="chapter" data-level="2.5" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#evaluating-the-model"><i class="fa fa-check"></i><b>2.5</b> Evaluating the Model</a></li>
<li class="chapter" data-level="2.6" data-path="implementing-neural-networks-with-keras.html"><a href="implementing-neural-networks-with-keras.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html"><i class="fa fa-check"></i><b>3</b> Data Preprocessing for Bioinformatics</a>
<ul>
<li class="chapter" data-level="3.1" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#encoding-categorical-data"><i class="fa fa-check"></i><b>3.1</b> Encoding Categorical Data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#one-hot-encoding-for-dna-sequences"><i class="fa fa-check"></i><b>3.1.1</b> One-Hot Encoding for DNA Sequences</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#handling-variable-length-sequences"><i class="fa fa-check"></i><b>3.2</b> Handling Variable-Length Sequences</a></li>
<li class="chapter" data-level="3.3" data-path="data-preprocessing-for-bioinformatics.html"><a href="data-preprocessing-for-bioinformatics.html#data-generators"><i class="fa fa-check"></i><b>3.3</b> Data Generators</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html"><i class="fa fa-check"></i><b>4</b> Convolutional Neural Networks for Genomics</a>
<ul>
<li class="chapter" data-level="4.1" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#the-convolutional-layer"><i class="fa fa-check"></i><b>4.1</b> The Convolutional Layer</a></li>
<li class="chapter" data-level="4.2" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#case-study-basset"><i class="fa fa-check"></i><b>4.2</b> Case Study: Basset</a></li>
<li class="chapter" data-level="4.3" data-path="convolutional-neural-networks-for-genomics.html"><a href="convolutional-neural-networks-for-genomics.html#exercise-build-a-tf-binding-predictor"><i class="fa fa-check"></i><b>4.3</b> Exercise: Build a TF Binding Predictor</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html"><i class="fa fa-check"></i><b>5</b> Large Language Models in Bioinformatics</a>
<ul>
<li class="chapter" data-level="5.1" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#from-word-embeddings-to-biological-embeddings"><i class="fa fa-check"></i><b>5.1</b> From Word Embeddings to Biological Embeddings</a></li>
<li class="chapter" data-level="5.2" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#pre-trained-models-protbert-and-dnabert"><i class="fa fa-check"></i><b>5.2</b> Pre-trained Models: ProtBERT and DNABERT</a></li>
<li class="chapter" data-level="5.3" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#how-to-use-a-pre-trained-model"><i class="fa fa-check"></i><b>5.3</b> How to Use a Pre-trained Model</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#example-using-dnabert-for-promoter-prediction"><i class="fa fa-check"></i><b>5.3.1</b> Example: Using DNABERT for Promoter Prediction</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="large-language-models-in-bioinformatics.html"><a href="large-language-models-in-bioinformatics.html#potential-applications"><i class="fa fa-check"></i><b>5.4</b> Potential Applications</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html"><i class="fa fa-check"></i><b>6</b> Advanced Topics &amp; Best Practices</a>
<ul>
<li class="chapter" data-level="6.1" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#regularization-fighting-overfitting"><i class="fa fa-check"></i><b>6.1</b> Regularization: Fighting Overfitting</a></li>
<li class="chapter" data-level="6.2" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#hyperparameter-tuning"><i class="fa fa-check"></i><b>6.2</b> Hyperparameter Tuning</a></li>
<li class="chapter" data-level="6.3" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#interpretability-what-did-my-model-learn"><i class="fa fa-check"></i><b>6.3</b> Interpretability: What Did My Model Learn?</a></li>
<li class="chapter" data-level="6.4" data-path="advanced-topics-best-practices.html"><a href="advanced-topics-best-practices.html#resources-for-the-future"><i class="fa fa-check"></i><b>6.4</b> Resources for the Future</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html"><i class="fa fa-check"></i><b>7</b> References and Further Reading</a>
<ul>
<li class="chapter" data-level="7.1" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#books-and-papers"><i class="fa fa-check"></i><b>7.1</b> Books and Papers</a></li>
<li class="chapter" data-level="7.2" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#online-courses"><i class="fa fa-check"></i><b>7.2</b> Online Courses</a></li>
<li class="chapter" data-level="7.3" data-path="references-and-further-reading.html"><a href="references-and-further-reading.html#useful-links"><i class="fa fa-check"></i><b>7.3</b> Useful Links</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Deep Learning for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="large-language-models-in-bioinformatics" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Large Language Models in Bioinformatics<a href="large-language-models-in-bioinformatics.html#large-language-models-in-bioinformatics" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="from-word-embeddings-to-biological-embeddings" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> From Word Embeddings to Biological Embeddings<a href="large-language-models-in-bioinformatics.html#from-word-embeddings-to-biological-embeddings" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Large Language Models (LLMs) like BERT and GPT work by learning deep contextual representations of words. We can apply the same principle to biological “words”.</p>
<ul>
<li><strong>Word -&gt; k-mer:</strong> A short biological sequence (e.g., a 3-6 amino acid peptide or a 3-6 nucleotide k-mer).</li>
<li><strong>Sentence -&gt; Protein/DNA sequence:</strong> The entire sequence is a sentence made of k-mer words.</li>
</ul>
</div>
<div id="pre-trained-models-protbert-and-dnabert" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Pre-trained Models: ProtBERT and DNABERT<a href="large-language-models-in-bioinformatics.html#pre-trained-models-protbert-and-dnabert" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Researchers have trained BERT-like models on massive corpora of protein sequences or DNA, creating powerful, general-purpose feature extractors for biology.</p>
<ul>
<li><strong>ProtBERT:</strong> Trained on millions of protein sequences from UniProt.</li>
<li><strong>DNABERT:</strong> Trained on genomes from humans and other species.</li>
</ul>
</div>
<div id="how-to-use-a-pre-trained-model" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> How to Use a Pre-trained Model<a href="large-language-models-in-bioinformatics.html#how-to-use-a-pre-trained-model" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The typical workflow is <strong>transfer learning</strong>:</p>
<ol style="list-style-type: decimal">
<li><strong>Feature Extraction:</strong> Use the pre-trained model to convert your raw sequences into high-quality numerical features (embeddings). Then, train a simple classifier (e.g., SVM, Logistic Regression) on these features.</li>
<li><strong>Fine-tuning:</strong> Start with the pre-trained model and continue training it on your specific, smaller dataset. This adapts the model’s knowledge to your task.</li>
</ol>
<div id="example-using-dnabert-for-promoter-prediction" class="section level3 hasAnchor" number="5.3.1">
<h3><span class="header-section-number">5.3.1</span> Example: Using DNABERT for Promoter Prediction<a href="large-language-models-in-bioinformatics.html#example-using-dnabert-for-promoter-prediction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will use the <code>transformers</code> library by Hugging Face.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="large-language-models-in-bioinformatics.html#cb8-1" tabindex="-1"></a><span class="co"># First, install the library: pip install transformers</span></span>
<span id="cb8-2"><a href="large-language-models-in-bioinformatics.html#cb8-2" tabindex="-1"></a><span class="im">from</span> transformers <span class="im">import</span> AutoTokenizer, AutoModel</span>
<span id="cb8-3"><a href="large-language-models-in-bioinformatics.html#cb8-3" tabindex="-1"></a><span class="im">import</span> torch</span>
<span id="cb8-4"><a href="large-language-models-in-bioinformatics.html#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="large-language-models-in-bioinformatics.html#cb8-5" tabindex="-1"></a><span class="co"># Load the pre-trained DNABERT model and tokenizer</span></span>
<span id="cb8-6"><a href="large-language-models-in-bioinformatics.html#cb8-6" tabindex="-1"></a>model_name <span class="op">=</span> <span class="st">&quot;zhihan1996/DNABERT-2-117M&quot;</span></span>
<span id="cb8-7"><a href="large-language-models-in-bioinformatics.html#cb8-7" tabindex="-1"></a>tokenizer <span class="op">=</span> AutoTokenizer.from_pretrained(model_name)</span>
<span id="cb8-8"><a href="large-language-models-in-bioinformatics.html#cb8-8" tabindex="-1"></a>model <span class="op">=</span> AutoModel.from_pretrained(model_name)</span>
<span id="cb8-9"><a href="large-language-models-in-bioinformatics.html#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="large-language-models-in-bioinformatics.html#cb8-10" tabindex="-1"></a><span class="co"># Tokenize a DNA sequence</span></span>
<span id="cb8-11"><a href="large-language-models-in-bioinformatics.html#cb8-11" tabindex="-1"></a>sequence <span class="op">=</span> <span class="st">&quot;ATCGATCGATCGATCG&quot;</span></span>
<span id="cb8-12"><a href="large-language-models-in-bioinformatics.html#cb8-12" tabindex="-1"></a>inputs <span class="op">=</span> tokenizer(sequence, return_tensors<span class="op">=</span><span class="st">&quot;pt&quot;</span>) <span class="co"># pt for PyTorch tensors</span></span>
<span id="cb8-13"><a href="large-language-models-in-bioinformatics.html#cb8-13" tabindex="-1"></a></span>
<span id="cb8-14"><a href="large-language-models-in-bioinformatics.html#cb8-14" tabindex="-1"></a><span class="co"># Get the model&#39;s embeddings</span></span>
<span id="cb8-15"><a href="large-language-models-in-bioinformatics.html#cb8-15" tabindex="-1"></a><span class="cf">with</span> torch.no_grad():</span>
<span id="cb8-16"><a href="large-language-models-in-bioinformatics.html#cb8-16" tabindex="-1"></a>    outputs <span class="op">=</span> model(<span class="op">**</span>inputs)</span>
<span id="cb8-17"><a href="large-language-models-in-bioinformatics.html#cb8-17" tabindex="-1"></a>    embeddings <span class="op">=</span> outputs.last_hidden_state <span class="co"># This is the contextual embedding for each token</span></span>
<span id="cb8-18"><a href="large-language-models-in-bioinformatics.html#cb8-18" tabindex="-1"></a></span>
<span id="cb8-19"><a href="large-language-models-in-bioinformatics.html#cb8-19" tabindex="-1"></a><span class="co"># The [CLS] token&#39;s embedding is often used as a representation of the whole sequence</span></span>
<span id="cb8-20"><a href="large-language-models-in-bioinformatics.html#cb8-20" tabindex="-1"></a>whole_sequence_embedding <span class="op">=</span> embeddings[<span class="dv">0</span>, <span class="dv">0</span>, :]</span>
<span id="cb8-21"><a href="large-language-models-in-bioinformatics.html#cb8-21" tabindex="-1"></a></span>
<span id="cb8-22"><a href="large-language-models-in-bioinformatics.html#cb8-22" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Shape of sequence embeddings: </span><span class="sc">{</span>embeddings<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">&quot;</span>)</span>
<span id="cb8-23"><a href="large-language-models-in-bioinformatics.html#cb8-23" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Shape of whole sequence embedding: </span><span class="sc">{</span>whole_sequence_embedding<span class="sc">.</span>shape<span class="sc">}</span><span class="ss">&quot;</span>)</span></code></pre></div>
<p>You can now use <code>whole_sequence_embedding</code> as input to a classifier for your task (e.g., promoter vs. non-promoter).</p>
</div>
</div>
<div id="potential-applications" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Potential Applications<a href="large-language-models-in-bioinformatics.html#potential-applications" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><strong>Variant Effect Prediction:</strong> Feed the reference and alternate allele sequences to a model and see how the embedding changes.</li>
<li><strong>Protein Function Prediction:</strong> Use ProtBERT embeddings to predict Gene Ontology terms.</li>
<li><strong>Structure Prediction:</strong> Embeddings can be used as inputs to predict secondary structure.</li>
</ul>
<blockquote>
<p><strong>Note:</strong> Fine-tuning these models requires significant computational resources (GPUs). For this course, we focus on feature extraction.</p>
</blockquote>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="convolutional-neural-networks-for-genomics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="advanced-topics-best-practices.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/danymukesha/DeepLearningFundamentals/edit/BRANCH/05-LLMs-Bioinformatics.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["Intro_Deep_Learning_Bioinformatics.pdf", "Intro_Deep_Learning_Bioinformatics.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
